src/protocols/LoRaWAN/LoRaWAN.cpp

```diff
@@ -48,172 +44,208 @@ void LoRaWANNode::setCSMA(uint8_t backoffMax, uint8_t difsSlots, bool enableCSMA
+    return(RADIOLIB_ERR_CHECKSUM_MISMATCH);
+  }
+
+  // copy the whole buffer over
+  memcpy(this->bufferNonces, persistentBuffer, RADIOLIB_LORAWAN_NONCES_BUF_SIZE);
+
+  // revert to inactive as long as no session is restored
+  this->bufferNonces[RADIOLIB_LORAWAN_NONCES_ACTIVE] = (uint8_t)false;
+
+  return(RADIOLIB_ERR_NONE);
+}
+
+uint8_t* LoRaWANNode::getBufferSession() {
+  // update buffer contents
+  this->saveSession();
+  
+  return(this->bufferSession);
 }
 
-int16_t LoRaWANNode::restore() {
+int16_t LoRaWANNode::setBufferSession(uint8_t* persistentBuffer) {
+  if(this->isJoined()) {
+    return(RADIOLIB_ERR_NONE);
+  }
+
+  // if all the session keys are 0, ignore
+  uint8_t allZeroes[4*RADIOLIB_AES128_BLOCK_SIZE] = { 0 };
+  if(memcmp(allZeroes, persistentBuffer, 4*RADIOLIB_AES128_BLOCK_SIZE) == 0) {
+    return(RADIOLIB_ERR_NETWORK_NOT_JOINED);
+  }
+
+  // check integrity of the whole buffer (compare checksum to included checksum)
+  uint16_t checkSum = LoRaWANNode::checkSum16(persistentBuffer, RADIOLIB_LORAWAN_SESSION_BUF_SIZE - 2);
+  uint16_t signature = LoRaWANNode::ntoh<uint16_t>(&persistentBuffer[RADIOLIB_LORAWAN_SESSION_SIGNATURE]);
+  if(signature != checkSum) {
+    RADIOLIB_DEBUG_PROTOCOL_PRINTLN("Calculated checksum: %04X, expected: %04X", checkSum, signature);
+    return(RADIOLIB_ERR_CHECKSUM_MISMATCH);
+  }
+
+  // the Nonces buffer holds a checksum signature - compare this to the signature that is in the session buffer
+  uint16_t signatureNonces = LoRaWANNode::ntoh<uint16_t>(&this->bufferNonces[RADIOLIB_LORAWAN_NONCES_SIGNATURE]);
+  uint16_t signatureInSession = LoRaWANNode::ntoh<uint16_t>(&persistentBuffer[RADIOLIB_LORAWAN_SESSION_NONCES_SIGNATURE]);
+  if(signatureNonces != signatureInSession) {
+    RADIOLIB_DEBUG_PROTOCOL_PRINTLN("The supplied session buffer does not match the Nonces buffer");
+    return(RADIOLIB_ERR_CHECKSUM_MISMATCH);
+  }
+
+  // copy the whole buffer over
+  memcpy(this->bufferSession, persistentBuffer, RADIOLIB_LORAWAN_SESSION_BUF_SIZE);
+
+  // as both the Nonces and session are restored, revert to active session
+  this->bufferNonces[RADIOLIB_LORAWAN_NONCES_ACTIVE] = (uint8_t)true;
+
+  return(RADIOLIB_ERR_NONE);
+}
+
+int16_t LoRaWANNode::restore(uint16_t checkSum, uint16_t lwMode, uint8_t lwClass, uint8_t freqPlan) {
   // if already joined, ignore
   if(this->activeMode != RADIOLIB_LORAWAN_MODE_NONE) {
-    return(this->activeMode);
+    return(RADIOLIB_ERR_NONE);
   }
 
-  Module* mod = this->phyLayer->getMod();
+  bool isSameKeys = LoRaWANNode::ntoh<uint16_t>(&this->bufferNonces[RADIOLIB_LORAWAN_NONCES_CHECKSUM]) == checkSum;
+  bool isSameMode = LoRaWANNode::ntoh<uint16_t>(&this->bufferNonces[RADIOLIB_LORAWAN_NONCES_MODE]) == lwMode;
+  bool isSameClass = LoRaWANNode::ntoh<uint8_t>(&this->bufferNonces[RADIOLIB_LORAWAN_NONCES_CLASS]) == lwClass;
+  bool isSamePlan  = LoRaWANNode::ntoh<uint8_t>(&this->bufferNonces[RADIOLIB_LORAWAN_NONCES_PLAN]) == freqPlan;
```

Comment:
Appreciate the consistency, but there's no need to convert single byte from network endianity to host endianity ;)

EDIT: this is also present at a few more places.

Reply:
Mostly a matter of preference.. speed vs consistency indeed. But I interpret you have a preference for speed in this case ;)

Reply:
Actually, I'd say we leave it like this. If there is future reason to convert this into e.g. uint16, it is much more likely that one forgets to convert it properly. Unless you have reasons to do change it.

Reply:
That's a valid point, let's leave it as-is nad hopefully more future-proof.